addons/account/models/account_account.py

```diff
@@ -784,22 +784,32 @@ def copy_data(self, default=None):
         cache = defaultdict(set)
         for account, vals in zip(self, vals_list):
             match vals.get('company_ids'):
-                case [(Command.LINK, company_id, 0)] | [(Command.SET, 0, [company_id])] | [company_id] if isinstance(company_id, int):
-                    company = self.env['res.company'].browse(company_id)
+                case [(Command.LINK, company_id, *_)] if isinstance(company_id, int):
+                    companies = self.env['res.company'].browse(company_id)
+                case [(Command.SET, 0, company_ids)] | company_ids if all(isinstance(company_id, int) for company_id in company_ids):
+                    companies = self.env['res.company'].browse(company_ids)
                 case _:
-                    raise ValueError(_("You may only give an account a single company at creation."))
-            if 'code' not in default:
-                start_code = account.with_company(company).code or account.with_company(account.company_ids[0]).code
-                vals['code'] = account.with_company(company)._search_new_account_code(start_code, cache[company])
-                cache[company].add(vals['code'])
+                    raise ValueError(_(
+                        "This copy() only accepts company_ids values that are "
+                        "a single [Command.link()], a single [Command.set()], or a list of company [ids]."
+                    ))
+
+            if 'code_by_company' not in default and ('code' not in default or len(companies) > 1):
+                companies_to_get_new_account_codes = companies if 'code' not in default else companies[1:]
+                vals['code_by_company'] = {}
+                for company in companies_to_get_new_account_codes:
+                    start_code = account.with_company(company).code or account.with_company(account.company_ids[0]).code
```

Comment:
This could for example happen if the original account belonged, say, to a single company, but the `company_ids` in the `default` contained several companies. In that case, the account might not have a code defined for every company in the company_ids, in which case we just take one of the account codes defined for the account as a starting point.